task/build-image-index/0.1/build-image-index.yaml

```diff
@@ -7,29 +7,33 @@ metadata:
   annotations:
     tekton.dev/pipelines.minVersion: "0.12.1"
     tekton.dev/tags: "image-build, konflux"
-  name: build-image-manifest
+  name: build-image-index
 spec:
   description: |-
-    This takes existing images and stiches them together into a multi platform image.
+    This takes existing Image Manifests and combines them in an Image Index.
   params:
-  - description: Reference of the image buildah will produce.
-    name: IMAGE
+  - name: IMAGE
+    description: The target image and tag where the image will be pushed to.
     type: string
-  - default: "true"
+  - name: TLSVERIFY
     description: Verify the TLS on the registry endpoint (for push/pull to a non-TLS registry)
-    name: TLSVERIFY
     type: string
+    default: "true"
   - name: COMMIT_SHA
-    description: The image is built from this commit.
+    description: The commit the image is built from.
     type: string
     default: ""
   - name: IMAGES
-    description: List of images that are to be merged into the multi platform image
+    description: List of Image Manifests to be referenced by the Image Index
     type: array
-  - default: ""
-    description: Delete image tag after specified time. Empty means to keep the image tag. Time values could be something like 1h, 2d, 3w for hours, days, and weeks, respectively.
-    name: IMAGE_EXPIRES_AFTER
+  - name: IMAGE_EXPIRES_AFTER
+    description: Delete image tag after specified time resulting in garbage collection of the digest. Empty means to keep the image tag. Time values could be something like 1h, 2d, 3w for hours, days, and weeks, respectively.
     type: string
+    default: ""
+  - name: BUILD_IMAGE_INDEX
+    description: Create an OCI image index referencing all passed params.IMAGES. This allows us to always include this task for a consistent pipeline even if a user does not want to generate the image index. If the image index generation is skipped, the task will forward values for params.IMAGES[0] to results.IMAGE_*.
```

Comment:
What does this simplify compared to not including this task in a single-arch pipeline?

I'm concerned about adding very specific behavior to tasks completely in service of one specific pipeline if that behavior doesn't make sense on its own. Our tasks should make sense standalone, not just in the sample pipeline

Reply:
Many of the builds that use the pipelines will compile binaries which cannot be run by multiple host architectures. Without an Image Index, the host will not have the ability to selectively pull the image that will work on the system. By adding the image index task by default, we enable users who only want to build in-cluster to still be conformant to the [image-spec](https://github.com/opencontainers/image-spec/blob/main/image-index.md)

> `platform` object
> 
> This OPTIONAL property describes the minimum runtime requirements of the image. This property SHOULD be present if its target is platform-specific.

Reply:
I get why we would want the image-index task in a single-arch pipeline, but why do we want a parameter that makes the task not do the only thing it's meant to do?

Reply:
Because it would be too hard for users to add/remove the task? Because there's no good way in Tekton to skip a task and still have the pipeline function? :disappointed:

Reply:
Maybe a param that would make more sense would be something like

```
ALWAYS_BUILD_INDEX: Build an image index even if IMAGES is of length 1. Default true.
```

Reply:
All of the above. By including the task as running by default (as a pass-through) then there is consistency in the result that contains the image generated. If users skip the task then they would have to change the path to the result.

I like the proposed parameter change. It also makes sense to set the default to `true` in the task and just override this as `false` in the specific pipelines for backwards compatibility. As a side note, this functionality would also enable _all_ pipelines for OLM operators to share a common pipeline definition since the same matrix pipeline can be used for single-arch builds without an image index as well as multi-arch with one (after any clusters are upgraded to OCP 4.15 so that the single-element fix is available).

Reply:
Resolved in the latest commit.